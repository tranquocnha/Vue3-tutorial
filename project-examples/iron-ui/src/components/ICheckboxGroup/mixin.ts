import type { ComputedRef, InjectionKey, Ref } from 'vue';

export interface CheckboxGroupInjection {
    name: string;
    value: ComputedRef<unknown[]>;
    size: Ref<string>;
    color: Ref<string>;
    disabled: Ref<boolean>;
    readonly: Ref<boolean>;
    onChange: (value: any) => void;
    onBlur: (event: FocusEvent) => void;
}

export const CheckboxGroupKey = Symbol('CheckboxGroup') as InjectionKey<CheckboxGroupInjection>;
