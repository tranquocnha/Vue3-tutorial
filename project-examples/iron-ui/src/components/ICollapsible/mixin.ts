import type { InjectionKey, Ref } from 'vue';

export interface CollapsibleInjection {
    activeItems: Ref<string[]>;
    onItemClick: (id: string) => void;
}

export const CollapsibleKey = Symbol('Collapsible') as InjectionKey<CollapsibleInjection>;
