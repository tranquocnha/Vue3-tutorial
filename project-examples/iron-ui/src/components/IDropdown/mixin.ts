import type { InjectionKey, Ref } from 'vue';

export interface DropdownInjection {
    disabled: Ref<boolean>;
    onItemClick: (event: Event) => void;
}

export const DropdownKey = Symbol('Dropdown') as InjectionKey<DropdownInjection>;
