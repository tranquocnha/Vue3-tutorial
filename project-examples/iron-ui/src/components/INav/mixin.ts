import { InjectionKey } from 'vue';

export interface NavInjection {
    onItemClick: (event: Event) => void;
}

export const NavKey = Symbol('Nav') as InjectionKey<NavInjection>;
