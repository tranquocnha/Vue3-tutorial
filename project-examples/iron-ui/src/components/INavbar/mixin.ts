import { InjectionKey, Ref } from 'vue';

export interface NavbarInjection {
    collapsible: Ref<boolean>;
    open: Ref<boolean>;
    onItemClick: (event: Event) => void;
}

export const NavbarKey = Symbol('Navbar') as InjectionKey<NavbarInjection>;
