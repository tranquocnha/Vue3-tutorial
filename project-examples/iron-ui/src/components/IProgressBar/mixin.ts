import { InjectionKey, Ref } from 'vue';

export interface ProgressInjection {
    min: Ref<number | string>;
    max: Ref<number | string>;
}

export const ProgressKey = Symbol('Progress') as InjectionKey<ProgressInjection>;
