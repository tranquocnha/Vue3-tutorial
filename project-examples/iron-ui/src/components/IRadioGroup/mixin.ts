import type { InjectionKey, Ref } from 'vue';

export interface RadioGroupInjection {
    name: Ref<string>;
    value: Ref;
    size: Ref<string>;
    color: Ref<string>;
    disabled: Ref<boolean>;
    readonly: Ref<boolean>;
    onChange: (value: any) => void;
    onBlur: (event: FocusEvent) => void;
}

export const RadioGroupKey = Symbol('RadioGroup') as InjectionKey<RadioGroupInjection>;
