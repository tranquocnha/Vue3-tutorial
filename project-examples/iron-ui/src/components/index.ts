// Composition API

export { IAlert } from '@inkline/inkline/components/IAlert';
export { IBadge } from '@inkline/inkline/components/IBadge';
export { IBreadcrumb } from '@inkline/inkline/components/IBreadcrumb';
export { IBreadcrumbItem } from '@inkline/inkline/components/IBreadcrumbItem';
export { IButton } from '@inkline/inkline/components/IButton';
export { IButtonGroup } from '@inkline/inkline/components/IButtonGroup';
export { ICard } from '@inkline/inkline/components/ICard';
export { ICheckbox } from '@inkline/inkline/components/ICheckbox';
export { ICheckboxGroup } from '@inkline/inkline/components/ICheckboxGroup';
export { ICollapsible } from '@inkline/inkline/components/ICollapsible';
export { ICollapsibleItem } from '@inkline/inkline/components/ICollapsibleItem';
export { IColumn } from '@inkline/inkline/components/IColumn';
export { IContainer } from '@inkline/inkline/components/IContainer';
export { IDropdown } from '@inkline/inkline/components/IDropdown';
export { IDropdownDivider } from '@inkline/inkline/components/IDropdownDivider';
export { IDropdownItem } from '@inkline/inkline/components/IDropdownItem';
export { IExpandTransition } from '@inkline/inkline/components/IExpandTransition';
export { IForm } from '@inkline/inkline/components/IForm';
export { IFormGroup } from '@inkline/inkline/components/IFormGroup';
export { IFormError } from '@inkline/inkline/components/IFormError';
export { IFormLabel } from '@inkline/inkline/components/IFormLabel';
export { IHamburgerMenu } from '@inkline/inkline/components/IHamburgerMenu';
export { IHeader } from '@inkline/inkline/components/IHeader';
export { IIcon } from '@inkline/inkline/components/IIcon';
export { IInput } from '@inkline/inkline/components/IInput';
export { ILayout } from '@inkline/inkline/components/ILayout';
export { ILayoutAside } from '@inkline/inkline/components/ILayoutAside';
export { ILayoutContent } from '@inkline/inkline/components/ILayoutContent';
export { ILayoutFooter } from '@inkline/inkline/components/ILayoutFooter';
export { ILayoutHeader } from '@inkline/inkline/components/ILayoutHeader';
export { IListGroup } from '@inkline/inkline/components/IListGroup';
export { IListGroupItem } from '@inkline/inkline/components/IListGroupItem';
export { ILoader } from '@inkline/inkline/components/ILoader';
export { IMark } from '@inkline/inkline/components/IMark';
export { IMedia } from '@inkline/inkline/components/IMedia';
export { IModal } from '@inkline/inkline/components/IModal';
export { INav } from '@inkline/inkline/components/INav';
export { INavItem } from '@inkline/inkline/components/INavItem';
export { INavbar } from '@inkline/inkline/components/INavbar';
export { INavbarBrand } from '@inkline/inkline/components/INavbarBrand';
export { INavbarCollapsible } from '@inkline/inkline/components/INavbarCollapsible';
export { INumberInput } from '@inkline/inkline/components/INumberInput';
export { IPagination } from '@inkline/inkline/components/IPagination';
export { IPopover } from '@inkline/inkline/components/IPopover';
export { IProgress } from '@inkline/inkline/components/IProgress';
export { IProgressBar } from '@inkline/inkline/components/IProgressBar';
export { IRadio } from '@inkline/inkline/components/IRadio';
export { IRadioGroup } from '@inkline/inkline/components/IRadioGroup';
export { IRow } from '@inkline/inkline/components/IRow';
export { ISelect } from '@inkline/inkline/components/ISelect';
export { ISelectOption } from '@inkline/inkline/components/ISelectOption';
export { ISidebar } from '@inkline/inkline/components/ISidebar';
export { ITable } from '@inkline/inkline/components/ITable';
export { ITabs } from '@inkline/inkline/components/ITabs';
export { ITab } from '@inkline/inkline/components/ITab';
export { ITabTitle } from '@inkline/inkline/components/ITabTitle';
export { ITextarea } from '@inkline/inkline/components/ITextarea';
export { IToast } from '@inkline/inkline/components/IToast';
export { IToastContainer } from '@inkline/inkline/components/IToastContainer';
export { IToggle } from '@inkline/inkline/components/IToggle';
export { ITooltip } from '@inkline/inkline/components/ITooltip';
