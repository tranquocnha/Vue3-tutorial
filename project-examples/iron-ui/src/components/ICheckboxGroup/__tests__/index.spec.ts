import { fireEvent, render } from '@testing-library/vue';
import { ICheckbox, ICheckboxGroup } from '@inkline/inkline/components';
import { InklineKey } from '@inkline/inkline/plugin';
import { createInkline } from '@inkline/inkline/__tests__/utils';
import { FormKey } from '@inkline/inkline/components/IForm/mixin';
import { ref } from 'vue';

describe('Components', () => {
    describe('ICheckboxGroup', () => {
        const props = {
            name: 'checkbox-group',
            color: 'light',
            size: 'md'
        };

        const stubs = {
            'i-checkbox': ICheckbox
        };

        const slots = {
            default: [
                '<i-checkbox color="light" name="checkbox-1" value="1" />',
                '<i-checkbox color="light" name="checkbox-2" value="2" />',
                '<i-checkbox color="light" name="checkbox-3" value="3" />'
            ]
        };

        it('should be named correctly', () => {
            expect(ICheckboxGroup.name).toEqual('ICheckboxGroup');
        });

        it('should render correctly', () => {
            const wrapper = render(ICheckboxGroup, {
                global: {
                    stubs,
                    provide: {
                        [InklineKey as symbol]: createInkline()
                    }
                },
                props,
                slots
            });
            expect(wrapper.html()).toMatchSnapshot();
        });

        describe('props', () => {
            describe('name', () => {
                it('should have randomly generated name uid', async () => {
                    const wrapper = render(ICheckboxGroup, {
                        global: {
                            stubs,
                            provide: {
                                [InklineKey as symbol]: createInkline()
                            }
                        },
                        props: {
                            color: props.color,
                            size: props.size
                        },
                        slots
                    });

                    expect(wrapper.container.firstChild).toHaveAttribute(
                        'name',
                        expect.stringContaining('checkbox-group')
                    );
                });
            });
        });

        describe('computed', () => {
            describe('classes', () => {
                it('should add classes based on props', () => {
                    const wrapper = render(ICheckboxGroup, {
                        global: {
                            stubs,
                            provide: {
                                [InklineKey as symbol]: createInkline()
                            }
                        },
                        props: {
                            disabled: true,
                            readonly: true,
                            inline: true,
                            ...props
                        },
                        slots
                    });

                    expect(wrapper.container.firstChild).toHaveClass(
                        `-${props.color}`,
                        `-${props.size}`,
                        '-disabled',
                        '-readonly',
                        '-inline'
                    );
                });
            });

            describe('checked', () => {
                it('should be equal to schema.value if schema', async () => {
                    const value = ['1'];
                    const onBlur = vi.fn();
                    const onInput = vi.fn();
                    const wrapper = render(ICheckboxGroup, {
                        global: {
                            stubs,
                            provide: {
                                [InklineKey as symbol]: createInkline(),
                                [FormKey as symbol]: {
                                    disabled: ref(false),
                                    readonly: ref(false),
                                    schema: ref({
                                        [props.name]: {
                                            value
                                        }
                                    }),
                                    onBlur,
                                    onInput
                                }
                            }
                        },
                        props: {
                            value,
                            ...props
                        },
                        slots
                    });
                    const checkboxes = await wrapper.getAllByRole('checkbox');

                    expect(checkboxes[0].querySelector('input')).toBeChecked();
                });

                it('should be equal to modelValue', async () => {
                    const modelValue = ['1', '2', '3'];
                    const wrapper = render(ICheckboxGroup, {
                        global: {
                            stubs,
                            provide: {
                                [InklineKey as symbol]: createInkline()
                            }
                        },
                        props: {
                            modelValue,
                            ...props
                        },
                        slots
                    });

                    const checkboxes = wrapper.container.querySelectorAll('input');

                    expect(checkboxes[0]).toBeChecked();
                    expect(checkboxes[1]).toBeChecked();
                    expect(checkboxes[2]).toBeChecked();
                });
            });
        });

        describe('methods', () => {
            describe('onChange', () => {
                it('should update modelValue when checking checkbox', async () => {
                    const wrapper = render(ICheckboxGroup, {
                        global: {
                            stubs,
                            provide: {
                                [InklineKey as symbol]: createInkline()
                            }
                        },
                        props: {
                            modelValue: [],
                            ...props
                        },
                        slots
                    });
                    const checkboxes = wrapper.container.querySelectorAll('label');

                    await fireEvent.click(checkboxes[0]);

                    expect(wrapper.emitted()['update:modelValue'][0]).toEqual([['1']]);
                });

                it('should update modelValue when unchecking checkbox', async () => {
                    const wrapper = render(ICheckboxGroup, {
                        global: {
                            stubs,
                            provide: {
                                [InklineKey as symbol]: createInkline()
                            }
                        },
                        props: {
                            modelValue: ['1'],
                            ...props
                        },
                        slots
                    });
                    const checkboxes = wrapper.container.querySelectorAll('label');

                    await fireEvent.click(checkboxes[0]);

                    expect(wrapper.emitted()['update:modelValue'][0]).toEqual([[]]);
                });

                it('should call parent form onInput when checking checkbox', async () => {
                    const onBlur = vi.fn();
                    const onInput = vi.fn();
                    const wrapper = render(ICheckboxGroup, {
                        global: {
                            stubs,
                            provide: {
                                [InklineKey as symbol]: createInkline(),
                                [FormKey as symbol]: {
                                    disabled: ref(false),
                                    readonly: ref(false),
                                    onBlur,
                                    onInput
                                }
                            }
                        },
                        props: {
                            modelValue: [],
                            ...props
                        },
                        slots
                    });
                    const checkboxes = wrapper.container.querySelectorAll('label');

                    await fireEvent.click(checkboxes[0]);

                    expect(onInput).toHaveBeenCalled();
                });
            });
        });
    });
});
