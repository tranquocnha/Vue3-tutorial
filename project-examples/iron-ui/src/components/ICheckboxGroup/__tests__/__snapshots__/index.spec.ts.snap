// Vitest Snapshot v1

exports[`Components > ICheckboxGroup > should render correctly 1`] = `
"<div class=\\"form-group checkbox-group -light -md -disabled -readonly\\" name=\\"checkbox-group\\" role=\\"checkboxgroup\\">
  <!-- @slot default Slot for default checkbox group options -->
  <div class=\\"checkbox -light -md\\" aria-checked=\\"false\\" role=\\"checkbox\\"><input type=\\"checkbox\\" name=\\"checkbox-1\\" aria-hidden=\\"true\\" aria-checked=\\"false\\"><label class=\\"checkbox-label\\" tabindex=\\"0\\">
      <!-- @slot default Slot for default checkbox label -->
    </label></div>
  <div class=\\"checkbox -light -md\\" aria-checked=\\"false\\" role=\\"checkbox\\"><input type=\\"checkbox\\" name=\\"checkbox-2\\" aria-hidden=\\"true\\" aria-checked=\\"false\\"><label class=\\"checkbox-label\\" tabindex=\\"0\\">
      <!-- @slot default Slot for default checkbox label -->
    </label></div>
  <div class=\\"checkbox -light -md\\" aria-checked=\\"false\\" role=\\"checkbox\\"><input type=\\"checkbox\\" name=\\"checkbox-3\\" aria-hidden=\\"true\\" aria-checked=\\"false\\"><label class=\\"checkbox-label\\" tabindex=\\"0\\">
      <!-- @slot default Slot for default checkbox label -->
    </label></div>
</div>"
`;
